components/nexusoperations/completion.go

```diff
@@ -196,17 +181,21 @@ func CompletionHandler(
 		if err := fabricateStartedEventIfMissing(node, requestID, operationID, startTime, links); err != nil {
 			return err
 		}
-		err := hsm.MachineTransition(node, func(operation Operation) (hsm.TransitionOutput, error) {
-			if requestID != "" && operation.RequestId != requestID {
-				isRetryableNotFoundErr = false
-				return hsm.TransitionOutput{}, serviceerror.NewNotFound("operation not found")
-			}
-			if opFailedError != nil {
-				return handleUnsuccessfulOperationError(node, operation, opFailedError, CompletionSourceCallback)
-			}
-			return handleSuccessfulOperationResult(node, operation, result, nil, CompletionSourceCallback)
-		})
-		// TODO(bergundy): Remove this once the operation auto-deletes itself from the tree on completion.
+		operation, err := hsm.MachineData[Operation](node)
+		if err != nil {
+			return nil
+		}
+		if requestID != "" && operation.RequestId != requestID {
+			isRetryableNotFoundErr = false
+			return serviceerror.NewNotFound("operation not found")
+		}
+		if opFailedError != nil {
+			err = handleUnsuccessfulOperationError(node, operation, opFailedError)
+		} else {
+			err = handleSuccessfulOperationResult(node, operation, result, nil)
```

Comment:
do you need to pass in links?

Reply:
No that happens only in the start event.